### PR TITLE
Add support for Tensor Cores in the BLAS bindings.

### PR DESCRIPTION
**Pull request by abergeron**

### CODE REVIEW

File: src/gpuarray_blas_cuda_cublas.c [hunk 1]
```diff
@@ -352,11 +368,23 @@ static int sgemm(cb_order order, cb_transpose transA, cb_transpose transB,
   GA_CUDA_EXIT_ON_ERROR(ctx, cuda_wait(B, CUDA_WAIT_READ));
   GA_CUDA_EXIT_ON_ERROR(ctx, cuda_wait(C, CUDA_WAIT_ALL));
 
-  CUBLAS_EXIT_ON_ERROR(ctx, cublasSgemm(h->h,
-                                        convT(transA), convT(transB), M, N, K,
-                                        &alpha, ((float *)A->ptr) + offA, lda,
-                                        ((float *)B->ptr) + offB, ldb, &beta,
-                                        ((float *)C->ptr) + offC, ldc));
+  if (cublasGemmEx != NULL && h->tensorCore) {
```
Comment (nouiz) [inline src/gpuarray_blas_cuda_cublas.c]: I would not use tensor core with sgemm as this downcast the input to float16 before computation with float32 output.

I don't have an idea of how to add tensorcore support for float32 in the user interface. I would let this for later if this get needed. Do you have an idea of how to add this easily?

**Comment by abergeron** [comment]
Do not merge this yet, more work is incoming.

File: src/gpuarray_blas_cuda_cublas.c (hunk 1 above)
Comment (abergeron) [reply src/gpuarray_blas_cuda_cublas.c]: I would just disable this part.  If there is demand for this, then we'll see.

Comment (nouiz) [reply src/gpuarray_blas_cuda_cublas.c]: Normally, you don't like dead code:) But if you think this is better, I'm fine with such few dead code.

Comment (abergeron) [reply src/gpuarray_blas_cuda_cublas.c]: By disable I mean delete :)

**Comment by abergeron** [comment]
Ok tests are weird, but the code should not be attempting to use tensor core in float32 anymore, while still using them for float16.

**Comment by nouiz** [comment]
I have this error on a GT750 and V100:

```
======================================================================
ERROR: theano.gpuarray.tests.test_blas.test_float16
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Tmp/lisa/os_v5/anaconda/lib/python2.7/site-packages/nose/case.py", line
 197, in runTest
    self.test(*self.arg)
  File "/u/bastienf/repos/Theano/theano/gpuarray/tests/test_blas.py", line 56, i
n test_float16
    out = f()
  File "/u/bastienf/repos/Theano/theano/compile/function_module.py", line 917, i
n __call__
    storage_map=getattr(self.fn, 'storage_map', None))
  File "/u/bastienf/repos/Theano/theano/gof/link.py", line 325, in raise_with_op
    reraise(exc_type, exc_value, exc_trace)
  File "/u/bastienf/repos/Theano/theano/compile/function_module.py", line 903, i
n __call__
    self.fn() if output_subset is None else\
  File "pygpu/blas.pyx", line 47, in pygpu.blas.pygpu_blas_rgemm (pygpu/blas.c:2
028)
GpuArrayException: ('cublasGemmEx(h->h, convT(transA), convT(transB), M, N, K, &
alpha, ((uint16_t *)A->ptr) + offA, CUDA_R_16F, lda, ((uint16_t *)B->ptr) + offB
, CUDA_R_16F, ldb, &beta, ((uint16_t *)C->ptr) + offC, CUDA_R_16F, ldc, CUDA_R_3
2F, CUBLAS_GEMM_DFALT_TENSOR_OP): (cublas) Invalid value.', 11)
Apply node that caused the error: GpuGemm{inplace=False}(InplaceGpuDimShuffle{0,
x}.0, HostFromGpu(gpuarray).0, <GpuArrayType<None>(float16, matrix)>, InplaceGpu
DimShuffle{0,x}.0, HostFromGpu(gpuarray).0)
Toposort index: 4
Inputs types: [GpuArrayType<None>(float16, col), TensorType(float32, scalar), Gp
uArrayType<None>(float16, matrix), GpuArrayType<None>(float16, col), TensorType(
float32, scalar)]
Inputs shapes: [(3, 1), (), (3, 3), (3, 1), ()]
Inputs strides: [(2, 2), (), (6, 2), (2, 2), ()]
Inputs values: [gpuarray.array([[ 0.86865234],
       [ 0.38427734],
       [ 0.78564453]], dtype=float16), array(1.0, dtype=float32), 'not shown', g
puarray.array([[-0.1385498 ],
       [ 0.38525391],
       [-0.01589966]], dtype=float16), array(0.5, dtype=float32)]
Outputs clients: [[InplaceGpuDimShuffle{0}(GpuGemm{inplace=False}.0)]]
```

I did a full delete of the old libgpuarray (the build directories and installed files)

**Comment by abergeron** [comment]
It works for me in float 16 and cuda9 on a 750, so this is weird.

**Comment by abergeron** [comment]
Make sure you don't have a mixed environment since cublas 8.0 actually gives that error and may get picked up because of other modules in the theano cache.

**Comment by nouiz** [comment]
On the V100 computer, there is only cuda9.

**Comment by abergeron** [comment]
Apparently requesting tensor op errors out on some non-volta cards.  Since it has no chance of working anyway, let's just not try.  This is what the last commit does.

**Comment by nouiz** [comment]
On a GTX 750 with cuda 8, I have this error:

```
$ theano-nose theano/gpuarray/tests/test_blas.py -s --pdb --pdb-failure
Using cuDNN version 7002 on context None
Mapped name None to device cuda: GeForce GTX 750 (0000:05:00.0)
.Disabling C code for BatchedDot due to unsupported float16
> /u/bastienf/repos/Theano/theano/compile/function_module.py(903)__call__()
-> self.fn() if output_subset is None else\
(Pdb) l
912  	                if hasattr(self.fn, 'thunks'):
913  	                    thunk = self.fn.thunks[self.fn.position_of_error]
914  	                gof.link.raise_with_op(
915  	                    node=self.fn.nodes[self.fn.position_of_error],
916  	                    thunk=thunk,
917  ->	                    storage_map=getattr(self.fn, 'storage_map', None))
918  	            else:
919  	                # old-style linkers raise their own exceptions
920  	                raise
921  	
922  	        dt_fn = time.time() - t0_fn
(Pdb) s
E...........................................SSS................SSS...........
======================================================================
ERROR: test_all (theano.gpuarray.tests.test_blas.GpuGemmBatchTester)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/u/bastienf/repos/Theano/theano/gpuarray/tests/test_basic_ops.py", line 97, in test_all
    self.run_case(testname, inputs)
  File "/u/bastienf/repos/Theano/theano/gpuarray/tests/test_basic_ops.py", line 131, in run_case
    variables = f_tst()
  File "/u/bastienf/repos/Theano/theano/compile/function_module.py", line 917, in __call__
    storage_map=getattr(self.fn, 'storage_map', None))
  File "/u/bastienf/repos/Theano/theano/gof/link.py", line 325, in raise_with_op
    reraise(exc_type, exc_value, exc_trace)
  File "/u/bastienf/repos/Theano/theano/compile/function_module.py", line 903, in __call__
    self.fn() if output_subset is None else\
RuntimeError: ('Blas operation not supported by library in use: hgemmBatch\nApply node that caused the error: GpuGemmBatch{inplace=True}(GpuAllocEmpty{dtype=\'float16\', context_name=None}.0, TensorConstant{1.0}, GpuFromHost<None>.0, GpuFromHost<None>.0, TensorConstant{0.0})\nToposort index: 11\nInputs types: [GpuArrayType<None>(float16, 3D), TensorType(float32, scalar), GpuArrayType<None>(float16, 3D), GpuArrayType<None>(float16, 3D), TensorType(float32, scalar)]\nInputs shapes: [(3, 4, 7), (), (3, 4, 4), (3, 4, 7), ()]\nInputs strides: [(56, 14, 2), (), (32, 8, 2), (56, 14, 2), ()]\nInputs values: [\'not shown\', array(1.0, dtype=float32), \'not shown\', \'not shown\', array(0.0, dtype=float32)]\nOutputs clients: [[GpuElemwise{Composite{((i0 * i1) + (i2 * i3))}}[(0, 1)]<gpuarray>(InplaceGpuDimShuffle{x,x,x}.0, GpuGemmBatch{inplace=True}.0, InplaceGpuDimShuffle{x,x,x}.0, GpuFromHost<None>.0)]]\n\nBacktrace when the node is created(use Theano flag traceback.limit=N to make it longer):\n  File "/u/bastienf/repos/Theano/theano/gpuarray/tests/test_basic_ops.py", line 97, in test_all\n    self.run_case(testname, inputs)\n  File "/u/bastienf/repos/Theano/theano/gpuarray/tests/test_basic_ops.py", line 115, in run_case\n    f_tst = inplace_func([], node_tst.outputs, mode=mode_gpu)\n  File "/u/bastienf/repos/Theano/theano/gpuarray/tests/test_basic_ops.py", line 42, in inplace_func\n    on_unused_input=on_unused_input, name=name)\n  File "/u/bastienf/repos/Theano/theano/compile/function.py", line 317, in function\n    output_keys=output_keys)\n  File "/u/bastienf/repos/Theano/theano/compile/pfunc.py", line 486, in pfunc\n    output_keys=output_keys)\n  File "/u/bastienf/repos/Theano/theano/compile/function_module.py", line 1839, in orig_function\n    name=name)\n  File "/u/bastienf/repos/Theano/theano/compile/function_module.py", line 1519, in __init__\n    optimizer_profile = optimizer(fgraph)\n  File "/u/bastienf/repos/Theano/theano/gof/opt.py", line 99, in __call__\n    return self.optimize(fgraph)\n  File "/u/bastienf/repos/Theano/theano/gof/opt.py", line 88, in optimize\n    ret = self.apply(fgraph, *args, **kwargs)\n  File "/u/bastienf/repos/Theano/theano/gof/opt.py", line 242, in apply\n    sub_prof = optimizer.optimize(fgraph)\n  File "/u/bastienf/repos/Theano/theano/gof/opt.py", line 88, in optimize\n    ret = self.apply(fgraph, *args, **kwargs)\n  File "/u/bastienf/repos/Theano/theano/gof/opt.py", line 2128, in apply\n    nb += self.process_node(fgraph, node)\n  File "/u/bastienf/repos/Theano/theano/gof/opt.py", line 2019, in process_node\n    replacements = lopt.transform(node)\n  File "/u/bastienf/repos/Theano/theano/gof/opt.py", line 1366, in transform\n    new_repl = opt.transform(node)\n  File "/u/bastienf/repos/Theano/theano/gpuarray/opt_util.py", line 333, in opt\n    return maker(node, inputs)\n  File "/u/bastienf/repos/Theano/theano/gpuarray/blas.py", line 1871, in local_inplace_gpuagemmbatch\n    return [gpugemmbatch_inplace(*inputs)]\n  File "/u/bastienf/repos/Theano/theano/gof/op.py", line 615, in __call__\n    node = self.make_node(*inputs, **kwargs)\n  File "/u/bastienf/repos/Theano/theano/gpuarray/blas.py", line 394, in make_node\n    return Apply(self, [C, alpha, A, B, beta], [C.type()])\n  File "/u/bastienf/repos/Theano/theano/gof/type.py", line 419, in __call__\n    return utils.add_tag_trace(self.make_variable(name))\n  File "/u/bastienf/repos/Theano/theano/gof/utils.py", line 108, in add_tag_trace\n    tr = simple_extract_stack(limit=user_line, skips=skips)\n  File "/Tmp/lisa/os_v5/anaconda/lib/python2.7/unittest/case.py", line 329, in run\n    testMethod()\n  File "/u/bastienf/repos/Theano/theano/gpuarray/tests/test_basic_ops.py", line 97, in test_all\n    self.run_case(testname, inputs)\n  File "/u/bastienf/repos/Theano/theano/gpuarray/tests/test_basic_ops.py", line 115, in run_case\n    f_tst = inplace_func([], node_tst.outputs, mode=mode_gpu)\n  File "/u/bastienf/repos/Theano/theano/gpuarray/tests/test_basic_ops.py", line 42, in inplace_func\n    on_unused_input=on_unused_input, name=name)\n  File "/u/bastienf/repos/Theano/theano/compile/function.py", line 317, in function\n    output_keys=output_keys)\n  File "/u/bastienf/repos/Theano/theano/compile/pfunc.py", line 486, in pfunc\n    output_keys=output_keys)\n  File "/u/bastienf/repos/Theano/theano/compile/function_module.py", line 1839, in orig_function\n    name=name)\n  File "/u/bastienf/repos/Theano/theano/compile/function_module.py", line 1519, in __init__\n    optimizer_profile = optimizer(fgraph)\n  File "/u/bastienf/repos/Theano/theano/gof/opt.py", line 99, in __call__\n    return self.optimize(fgraph)\n  File "/u/bastienf/repos/Theano/theano/gof/opt.py", line 88, in optimize\n    ret = self.apply(fgraph, *args, **kwargs)\n  File "/u/bastienf/repos/Theano/theano/gof/opt.py", line 242, in apply\n    sub_prof = optimizer.optimize(fgraph)\n  File "/u/bastienf/repos/Theano/theano/gof/opt.py", line 88, in optimize\n    ret = self.apply(fgraph, *args, **kwargs)\n  File "/u/bastienf/repos/Theano/theano/gof/opt.py", line 242, in apply\n    sub_prof = optimizer.optimize(fgraph)\n  File "/u/bastienf/repos/Theano/theano/gof/opt.py", line 88, in optimize\n    ret = self.apply(fgraph, *args, **kwargs)\n  File "/u/bastienf/repos/Theano/theano/gpuarray/opt.py", line 417, in apply\n    node.outputs)\n  File "/u/bastienf/repos/Theano/theano/gpuarray/opt.py", line 1349, in local_gpua_gemmbatch\n    a, b, np.asarray(0.0, dtype=out_dtype))\n  File "/u/bastienf/repos/Theano/theano/gof/op.py", line 615, in __call__\n    node = self.make_node(*inputs, **kwargs)\n  File "/u/bastienf/repos/Theano/theano/gpuarray/blas.py", line 394, in make_node\n    return Apply(self, [C, alpha, A, B, beta], [C.type()])\n  File "/u/bastienf/repos/Theano/theano/gof/type.py", line 419, in __call__\n    return utils.add_tag_trace(self.make_variable(name))\n  File "/u/bastienf/repos/Theano/theano/gof/utils.py", line 108, in add_tag_trace\n    tr = simple_extract_stack(limit=user_line, skips=skips)\n  File "/u/bastienf/repos/Theano/theano/gpuarray/tests/test_basic_ops.py", line 105, in run_case\n    node_tst = safe_make_node(self.op, *inputs_tst)\n  File "/u/bastienf/repos/Theano/theano/tensor/tests/test_basic.py", line 203, in safe_make_node\n    node = op(*inputs)\n  File "/u/bastienf/repos/Theano/theano/gpuarray/tests/test_blas.py", line 153, in <lambda>\n    op=lambda z, alpha, x, y, beta: alpha * batched_dot(x, y) + beta * z,\n  File "/u/bastienf/repos/Theano/theano/gof/op.py", line 615, in __call__\n    node = self.make_node(*inputs, **kwargs)\n  File "/u/bastienf/repos/Theano/theano/tensor/blas.py", line 2036, in make_node\n    return Apply(self, upcasted_inputs, [T.tensor(dtype, broadcastable)])\n  File "/u/bastienf/repos/Theano/theano/tensor/basic.py", line 641, in tensor\n    return TensorType(*args, **kwargs)(name=name)\n  File "/u/bastienf/repos/Theano/theano/gof/type.py", line 419, in __call__\n    return utils.add_tag_trace(self.make_variable(name))\n  File "/u/bastienf/repos/Theano/theano/gof/utils.py", line 108, in add_tag_trace\n    tr = simple_extract_stack(limit=user_line, skips=skips)\n  File "/u/bastienf/repos/Theano/theano/gpuarray/tests/test_basic_ops.py", line 105, in run_case\n    node_tst = safe_make_node(self.op, *inputs_tst)\n  File "/u/bastienf/repos/Theano/theano/tensor/tests/test_basic.py", line 203, in safe_make_node\n    node = op(*inputs)\n  File "/u/bastienf/repos/Theano/theano/gpuarray/tests/test_blas.py", line 153, in <lambda>\n    op=lambda z, alpha, x, y, beta: alpha * batched_dot(x, y) + beta * z,\n  File "/u/bastienf/repos/Theano/theano/gof/op.py", line 615, in __call__\n    node = self.make_node(*inputs, **kwargs)\n  File "/u/bastienf/repos/Theano/theano/tensor/blas.py", line 2036, in make_node\n    return Apply(self, upcasted_inputs, [T.tensor(dtype, broadcastable)])\n  File "/u/bastienf/repos/Theano/theano/tensor/basic.py", line 641, in tensor\n    return TensorType(*args, **kwargs)(name=name)\n  File "/u/bastienf/repos/Theano/theano/gof/type.py", line 419, in __call__\n    return utils.add_tag_trace(self.make_variable(name))\n  File "/u/bastienf/repos/Theano/theano/gof/utils.py", line 108, in add_tag_trace\n    tr = simple_extract_stack(limit=user_line, skips=skips)\n\nHINT: Use the Theano flag \'exception_verbosity=high\' for a debugprint and storage map footprint of this apply node.', 'Test GpuGemmBatch{inplace=True}::float16: exception when calling the Function')

----------------------------------------------------------------------
Ran 78 tests in 71.386s

FAILED (SKIP=6, errors=1)

```

**Comment by nouiz** [comment]
The error isn't related to this PR. @abergeron will investigate.

So merging.

**Pull request merged**